cassandra-analytics-core/src/test/java/org/apache/cassandra/spark/bulkwriter/cloudstorage/coordinated/MultiClusterContainerTest.java

```diff
@@ -0,0 +1,161 @@
+        .isExactlyInstanceOf(IllegalStateException.class)
+        .hasMessage("Cannot set value for null cluster when the container is used for coordinated-write");
+    }
+
+    @Test
+    void testUpdateValue()
+    {
+        for (String clusterId : Arrays.asList(null, "cluster1"))
+        {
+            MultiClusterContainer<Value> container = new MultiClusterContainer<>();
+            container.updateValue(clusterId, v -> {
+                assertThat(v).isNull();
+                return null;
+            });
+
+            container.setValue(clusterId, new Value());
+            assertThat(container.getValueOrThrow(clusterId).a).isEqualTo(1);
+            container.updateValue(clusterId, v -> {
+                v.a = 100;
+                return v;
+            });
+            assertThat(container.getValueOrThrow(clusterId).a).isEqualTo(100);
+
+            // remove the value
+            container.updateValue(clusterId, v -> null);
+            assertThat(container.getValueOrNull(clusterId)).isNull();
+        }
+    }
+
+    @Test
+    void testHashCodeAndEquals()
+    {
+        Value value = new Value();
+        MultiClusterContainer<Value> container1 = new MultiClusterContainer<>();
+        container1.setValue(null, value);
+        MultiClusterContainer<Value> container2 = new MultiClusterContainer<>();
+        container2.setValue(null, value);
+        assertThat(container1.hashCode()).isEqualTo(container2.hashCode());
+        assertThat(container1).isEqualTo(container2);
+
+        container1 = new MultiClusterContainer<>();
+        container1.setValue("cluster1", value);
+        container2 = new MultiClusterContainer<>();
+        container2.setValue("cluster1", value);
+        assertThat(container1.hashCode()).isEqualTo(container2.hashCode());
+        assertThat(container1).isEqualTo(container2);
+
+        container1.setValue("cluster2", value);
+        assertThat(container1.hashCode()).isNotEqualTo(container2.hashCode());
+        assertThat(container1).isNotEqualTo(container2);
+    }
+
+    @Test
+    void testGetAnyValue()
+    {
+        for (String clusterId : Arrays.asList(null, "cluster1"))
+        {
+            MultiClusterContainer<Value> container = new MultiClusterContainer<>();
+            assertThat(container.getAnyValue()).isNull();
+
+            container.setValue(clusterId, new Value());
+            assertThat(container.getAnyValue()).isNotNull();
+        }
+    }
+
+    @Test
+    void testSize()
+    {
+        MultiClusterContainer<Value> container = new MultiClusterContainer<>();
+        assertThat(container.size()).isZero();
+
+        container.setValue("cluster1", new Value());
+        assertThat(container.size()).isOne();
+
+        container.setValue("cluster2", new Value());
+        assertThat(container.size()).isEqualTo(2);
+
+        container = new MultiClusterContainer<>();
+        container.setValue(null, new Value());
```

Comment:
small nit, maybe set the value twice to ensure we always have one entry when we update the null key twice?
```suggestion
        container.setValue(null, new Value());
        container.setValue(null, new Value());
```